impl/src/metadata.rs

```diff
@@ -92,189 +246,203 @@ impl CargoMetadataFetcher {
     Ok(crate_dir)
   }
 
-  pub fn fetch_crate_src(
+  fn inject_binaries_into_workspace(
     &self,
-    registry_url: &str,
-    dir: &Path,
-    lockfile_dir: &Path,
-    name: &str,
-    version: &str,
-  ) -> Result<(PathBuf, CargoWorkspaceFiles)> {
-    let src_dir = self.clone_crate_src(registry_url, dir, name, version)?;
-
-    let toml_path: PathBuf = src_dir.join("Cargo.toml");
-    let lock_path_opt: PathBuf =
-      cargo_generate_lockfile(src_dir.as_path(), lockfile_dir, name, version)?;
-
-    match toml_path.exists() && lock_path_opt.exists() {
-      true => Ok((
-        src_dir,
-        CargoWorkspaceFiles {
-          toml_path,
-          lock_path_opt: Some(lock_path_opt),
-        },
-      )),
-      false => Err(anyhow!(
-        "Cargo files could not be found for '{}' ({})",
-        name,
-        version
-      )),
-    }
-  }
-}
+    binary_deps: Vec<String>,
+    root_toml: &Path,
+  ) -> Result<()> {
+    let mut manifest = {
+      let content = fs::read_to_string(root_toml)?;
+      cargo_toml::Manifest::from_str(content.as_str())?
+    };
 
-impl Default for CargoMetadataFetcher {
-  fn default() -> CargoMetadataFetcher {
-    CargoMetadataFetcher::new(SYSTEM_CARGO_BIN_PATH, true)
-  }
-}
+    let mut workspace = match manifest.workspace {
+      Some(workspace) => workspace,
+      None => cargo_toml::Workspace::default(),
+    };
 
-impl MetadataFetcher for CargoMetadataFetcher {
-  fn fetch_metadata(&self, files: &CargoWorkspaceFiles) -> Result<Metadata> {
-    assert!(files.toml_path.is_file());
-    assert!(files.lock_path_opt.as_ref().map_or(true, |p| p.is_file()));
+    for dep in binary_deps.iter() {
+      workspace.members.push(dep.to_string());
+    }
 
-    let tempdir = TempDir::new()?;
-
-    // UNWRAP: Guarded by function assertion
-    let cargo_dir = match self.use_tempdir {
-      false => files.toml_path.as_path().parent().unwrap(),
-      true => {
-        let dir_path = tempdir.path();
-        let new_toml_path = dir_path.join(files.toml_path.file_name().unwrap());
-        fs::copy(files.toml_path.as_path(), new_toml_path)?;
-        if let Some(lock_path) = files.lock_path_opt.as_ref() {
-          let new_lock_path = dir_path.join(lock_path.file_name().unwrap());
-          fs::copy(lock_path.as_path(), new_lock_path)?;
-        }
+    manifest.workspace = Some(workspace);
 
-        dir_path
-      },
+    // cargo_toml::Manifest cannot be serialized direcly.
+    // see: https://gitlab.com/crates.rs/cargo_toml/-/issues/3
+    let value = toml::Value::try_from(&manifest)?;
+    match std::fs::write(root_toml, toml::to_string(&value)?) {
+      Ok(_) => Ok(()),
+      Err(err) => Err(anyhow!("{:?}", err)),
+    }
+  }
+
+  fn fetch_crate_checksum(&self, name: &str, version: &str) -> Result<String> {
+    let index_path_is_url = url::Url::parse(&self.index_url).is_ok();
+    let crate_index_path = if index_path_is_url {
```

Comment:
I think this is the second time this review I"m seeing this code block.

Two things.

1. Should this be moved into a function?
2. Should we just be opening this index in `new` or something and holding onto it?

Reply:
Where is the second instance of this code block?